R/lint.R

```diff
@@ -561,6 +561,161 @@ checkstyle_output <- function(lints, filename = "lintr_results.xml") {
+              "name": "lintr",
+              "informationUri": "https://lintr.r-lib.org/",
+              "version": "2.0.1",
+              "rules": [
+                {
+                  "id": "trailing_whitespace_linter",
+                  "fullDescription": {
+                    "text": "Trailing whitespace is superfluous."
+                  },
+                  "defaultConfiguration": {
+                    "level": "note"
+                  }
+                }
+              ]
+            }
+          },
+          "results": [
+            {
+              "ruleId": "trailing_whitespace_linter",
+              "ruleIndex": 0,
+              "message": {
+                "text": "Trailing blank lines are superfluous."
+              },
+              "locations": [
+                {
+                  "physicalLocation": {
+                    "artifactLocation": {
+                      "uri": "TestFileFolder/hello.r",
+                      "uriBaseId": "ROOTPATH"
+                    },
+                    "region": {
+                      "startLine": 2,
+                      "startColumn": 22,
+                      "snippet": {
+                        "text": "print(Hello World!) "
+                      }
+                    }
+                  }
+                }
+              ]
+            }
+          ],
+          "columnKind": "utf16CodeUnits",
+          "originalUriBaseIds": {
+            "ROOTPATH": {
+              "uri": "file:///C:/repos/repototest/"
+            }
+          }
+        }
+      ]
+    }',
+    simplifyVector = TRUE,
+    simplifyDataFrame = FALSE,
+    simplifyMatrix = FALSE
+  )
+
+  # assign values
+  sarif$runs[[1L]]$results <- NULL
+  sarif$runs[[1L]]$tool$driver$rules <- NULL
+  sarif$runs[[1L]]$tool$driver$version <-
+    as.character(utils::packageVersion("lintr"))
+  sarif$runs[[1L]]$originalUriBaseIds$ROOTPATH$uri <- ""
+  rule_index_exists <- FALSE
+  root_path_uri <- gsub("\\\\", "/", package_path)
+
+  if (startsWith(root_path_uri, "/")) {
+    root_path_uri <- paste("file://", root_path_uri, sep = "")
+  } else {
+    root_path_uri <- paste("file:///", root_path_uri, sep = "")
+  }
+
+  if (!endsWith(root_path_uri, "/")) {
+    root_path_uri <- paste(root_path_uri, "/", sep = "")
+  }
+
+  sarif$runs[[1L]]$originalUriBaseIds$ROOTPATH$uri <- root_path_uri
+
+  # loop and assign result values
+  for (lint in lints) {
+    one_result <- list()
+
+    if (is.null(sarif$runs[[1L]]$tool$driver$rules)) {
+      rule_index_exists <- 0L
+    } else {
+      rule_index_exists <-
+        which(sapply(sarif$runs[[1L]]$tool$driver$rules,
+                     function(x) x$id == lint$linter))
+      if (length(rule_index_exists) == 0L ||
+          is.na(rule_index_exists[1L])) {
+        rule_index_exists <- 0L
+      }
+    }
+
+    if (rule_index_exists == 0L) {
+      new_rule <- list(
+        id = lint$linter,
+        fullDescription = list(text = lint$message),
```

Comment:
Note that  `lint$message` is usually a customized message that incorporates context information for the specific piece of code that was linted, e.g. `undesirable_function_linter()` generates a message naming the concrete function that was detected. Is that okay for SARIF?

Reply:
Yes, currently this field lint$message is used for both the each result's message and also rule description. Rule description is required field can not be empty. In each result's message user can see the correct message for that instance of the violation which is correct and what we want; the rule description should be static and not have any variable in the string. In the future if we have a separated property like lint$rule_description we should use it. Many the rules lint$message I see is static so this is a good starting version.

Reply:
Thanks for the explanation!